### PR TITLE
RMC-134 Ad Hoc Fulfilment Request Consistency

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
UAC/QID pairs can now be created through an endpoint in the case API service which then sends out an RM_UAC_CREATED event. Case processor needs to ingest this new event, create the UAC QID link and distribute a UAC_UPDATED message so that the new UAC/QID pair becomes actively linked to the case it was created for.

# What has changed
* Create new UAC created queue, bind to UAC created exchange
* Add message receiver for UAC created queue
* Ingest UAC created message, create UAC/QID link, emit UAC_UPDATED event, log the RM_UAC_CREATED event

# How to test?
Build and run this and the linked case API branch then run the linked acceptance test branch. This will test the UAC_UPDATED events are emitted as well as the on request questionnaire fulfilment feature which makes use of this.

# Links
https://trello.com/c/AOdYVaz0/1004-rmc-134b2-eventual-consistency-for-uac-qid-creation-in-case-processor-and-case-api-part-3-8
https://github.com/ONSdigital/census-rm-case-api/pull/21
https://github.com/ONSdigital/census-rm-acceptance-tests/pull/103

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[RMC-134(B2) eventual consistency for UAC/QID creation in case processor and case api (part 3) (8)](https://trello.com/c/AOdYVaz0/1004-rmc-134b2-eventual-consistency-for-uac-qid-creation-in-case-processor-and-case-api-part-3-8)

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public void logRefusalEvent(
     if (StringUtils.isEmpty(event.getTransactionId())) {
       loggedEvent.setEventTransactionId(UUID.randomUUID());
     } else {
-      loggedEvent.setEventTransactionId(UUID.fromString(event.getTransactionId()));
+      loggedEvent.setEventTransactionId(event.getTransactionId());
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: hurrah for duplication

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 2]
```diff
@@ -100,7 +100,7 @@ public void logFulfilmentRequestedEvent(
...
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: hurrah for duplication

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 3]
```diff
@@ -145,11 +145,16 @@ public void logEvent(
...
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: hurrah for duplication

File: src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java [hunk 4]
```diff
@@ -89,4 +103,28 @@ public PayloadDTO emitUacUpdatedEvent(UacQidLink uacQidLink, Case caze, boolean
 
     return payloadDTO;
   }
+
+  public void ingestUacCreatedEvent(ResponseManagementEvent uacCreatedEvent) {
+    Optional<Case> linkedCase =
+        caseRepository.findByCaseId(uacCreatedEvent.getPayload().getUacQidCreated().getCaseId());
+    if (linkedCase.isEmpty()) {
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java]: petty but some spacing would be nice above blocks etc

File: src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    // Check the Uac Qid Link is created
+    Optional<UacQidLink> actualUacQidLink =
+        uacQidLinkRepository.findByQid(uacCreatedEvent.getPayload().getUacQidCreated().getQid());
+    assertTrue(actualUacQidLink.isPresent());
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getUac(), actualUacQidLink.get().getUac());
+    assertEquals(linkedCase.getCaseId(), actualUacQidLink.get().getCaze().getCaseId());
+  }
+}
```
Comment (LukeLoze1) [inline src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java]: As a team we need to decide whether we're going to always test @Transactional or not.  Some do it, some don't.  It's good to check that a failure will behave as expected (not ack msg and keep trying until it passes),  there's a transactionIT class which suppresses logging.

File: src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java [hunk 6]
```diff
@@ -16,23 +20,31 @@
...
+
   @Mock RabbitTemplate rabbitTemplate;
 
   @Mock UacQidServiceClient uacQidServiceClient;
 
+  @Mock EventLogger eventLogger;
+
   @InjectMocks UacProcessor underTest;
 
+  public static final EasyRandom easyRandom = new EasyRandom();
```
Comment (LukeLoze1) [inline src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java]: petty, but does this need to public

File: src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java [hunk 7]
```diff
@@ -76,4 +88,94 @@ public void testEmitUacUpdatedEvent() {
...
+        uacCreatedEvent.getPayload().getUacQidCreated().getQid(),
+        uacQidLinkArgumentCaptor.getValue().getQid());
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getUac(),
+        uacQidLinkArgumentCaptor.getValue().getUac());
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getCaseId(),
+        uacQidLinkArgumentCaptor.getValue().getCaze().getCaseId());
+  }
 }
```
Comment (LukeLoze1) [inline src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java]: There should be some sad :( path tests.  Checking that a RunTimeException is thrown when linkedCase.isEmpty() for example.  There's plenty of example in the project.   e.g. testReceiptedQidNotFound.  Again this isn't consistent across even this project

File: src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java [hunk 8]
```diff
@@ -76,4 +88,94 @@ public void testEmitUacUpdatedEvent() {
...
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getQid(),
+        uacQidLinkArgumentCaptor.getValue().getQid());
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getUac(),
+        uacQidLinkArgumentCaptor.getValue().getUac());
+    assertEquals(
+        uacCreatedEvent.getPayload().getUacQidCreated().getCaseId(),
+        uacQidLinkArgumentCaptor.getValue().getCaze().getCaseId());
+  }
```
Comment (LukeLoze1) [inline src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java]: There should be some sad :( path tests. Checking that a RunTimeException is thrown when linkedCase.isEmpty() for example. There's plenty of example in the project. e.g. testReceiptedQidNotFound. Again this isn't consistent across even this project

File: src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java (hunk 5 above)
Comment (AdamHawtin) [reply src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java]: @LukeLoze1 If we're going to require them then we need a tech session or something, I was not even aware they are a thing till now.

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java (hunk 1 above)
Comment (AdamHawtin) [reply src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: @LukeLoze1 I'm not sure why these `log*Type*Event` methods exist, non of them appear to do anything specific to the event type in their name and they could be replaced with one or two versions of the general `logEvent` method (like I used in this PR to log the UAC created events). I thought that refactor was out of the scope of this change though.

File: src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java (hunk 6 above)
Comment (AdamHawtin) [reply src/test/java/uk/gov/ons/census/casesvc/service/UacProcessorTest.java]: It's actually not even used anymore so I'll get rid.

File: src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java (hunk 5 above)
Comment (LukeLoze1) [reply src/test/java/uk/gov/ons/census/casesvc/messaging/UacCreatedEventReceiverIT.java]: yeap, they're quite easy - but we should just be consistent

**Pull request merged**